File: sd_card.cpp
#include "sd_card.h"
#include "esphome/core/log.h"

namespace esphome {
namespace sd_card {

static const char *TAG = "sd_card";

void SDCardComponent::setup() {
  ESP_LOGI(TAG, "SD Card component setup()");
}

void SDCardComponent::loop() {
  // Placeholder loop
}

}  // namespace sd_card
}  // namespace esphome

File: sd_card.py
import esphome.codegen as cg
import esphome.config_validation as cv
from esphome.components import spi
from esphome.const import CONF_ID

sd_card_ns = cg.esphome_ns.namespace("sd_card")
SDCardComponent = sd_card_ns.class_("SDCardComponent", cg.Component, spi.SPIDevice)

CONFIG_SCHEMA = cv.Schema({
    cv.GenerateID(): cv.declare_id(SDCardComponent),
}).extend(spi.spi_device_schema())

async def to_code(config):
    var = cg.new_Pvariable(config[CONF_ID])
    await cg.register_component(var, config)
    await spi.register_spi_device(var, config)
